File: xp-api/src/models/experiment.ts
import { z } from "zod";
import { PrismaSelectEntity } from "../utils/db/select-entity.js";
import { selectUser, userSchema } from "./user.js";
import { tgGeoSchema } from "./tg-geo.js";
import { tgMediaGroupItemSchema } from "./tg-media-group-item.js";

export const experimentSchema = z.object({
  id: z.string(),
  user: userSchema,
  tgText: z.string().nullable(),
  tgPhotoId: z.string().nullable(),
  tgVideoId: z.string().nullable(),
  tgVoiceId: z.string().nullable(),
  tgDocumentId: z.string().nullable(),
  tgVideoNoteId: z.string().nullable(),
  tgGeo: tgGeoSchema.nullable(),
  tgMediaGroup: z.array(tgMediaGroupItemSchema),
  createdAt: z.date(),
  completeBy: z.date(),
  completedAt: z.date().nullable(),
  canceledAt: z.date().nullable(),
});

export type Experiment = z.infer<typeof experimentSchema>;

export const selectExperiment: PrismaSelectEntity<Experiment> = {
  canceledAt: true,
  completeBy: true,
  completedAt: true,
  createdAt: true,
  id: true,
  tgDocumentId: true,
  tgGeo: true,
  tgMediaGroup: true,
  tgPhotoId: true,
  tgText: true,
  tgVideoId: true,
  tgVideoNoteId: true,
  tgVoiceId: true,
  user: { select: selectUser },
};

File: bot/core/models/experiment.py
from typing import Optional
from pydantic import BaseModel
from .user import User
from .tg_geo import TgGeo
from .tg_media_group_item import TgMediaGroupItem
from datetime import datetime

class Experiment(BaseModel):
    id: str
    user: User
    tgText: Optional[str]
    tgPhotoId: Optional[str]
    tgVideoId: Optional[str]
    tgVoiceId: Optional[str]
    tgDocumentId: Optional[str]
    tgVideoNoteId: Optional[str]
    tgGeo: Optional[TgGeo]
    tgMediaGroup: list[TgMediaGroupItem]
    createdAt: datetime
    completeBy: datetime
    completedAt: Optional[datetime]
    canceledAt: Optional[datetime]
